src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ChannelFactoryBase.cs

```diff
@@ -186,5 +191,25 @@ protected override void OnEndClose(IAsyncResult result)
         {
             ChainedCloseAsyncResult.End(result);
         }
+
+        private async Task OnCloseAsyncInternal(TimeSpan timeout)
+        {
+            IChannel[] currentChannels = _channels.ToArray();
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            foreach (IChannel channel in currentChannels)
+            {
+                IAsyncCommunicationObject iAsyncChannel = channel as IAsyncCommunicationObject;
+                if (iAsyncChannel != null)
+                {
+                    await iAsyncChannel.CloseAsync(timeout);
+                }
+                else
+                {
+                    channel.Close(timeoutHelper.RemainingTime());
+                }
+            }
```

Comment:
It's a good observation and gets to one of the points behind #1088. Given our current inheritance chain, that channel.Close() call is dead code and left in place to handle the unlikely custom channel.  We have several other places where this "as" cast happens and offers the alternative (but dead) call path.  #1088 is intended to ensure we can enforce we can hard cast to IAsyncCommunicationObject and drop the dead code.